### ISSUE TITLE
Calling private methods on mocks

### ISSUE THREAD

**Issue by elazar** [mlively/Phake#212]
I'm not sure the [documentation](http://phake.readthedocs.org/en/2.1/mocks.html#calling-private-and-protected-methods-on-mocks) for this feature is very clear on the [intended use case](https://github.com/mlively/Phake/issues/158), particularly as it relates to `private` methods.

The code example from the documentation doesn't make the use case very obvious.

```php
class MyClass
{
    private function foo()
    {
    }

    private static function bar()
    {
    }
}

$mock = Phake::mock('MyClass');

Phake::makeVisible($mock)->foo();

Phake::makeStaticVisible($mock)->bar();

//Both calls below will STILL fail
$mock->foo();
$mock::bar();
```

Why would you need to directly invoke a `private` method on a mock object if it's stubbed?

The documentation suggests a potential reason:

> So if you intend on the original method being called and you aren’t using [Partial Mocks](http://phake.readthedocs.org/en/2.1/method-stubbing.html#partial-mocks), then you can just enable calling-the-parent for that method call using the `thenCallParent()` answer. 

However, if I'm understanding correctly what this is suggesting...

```php
class PrivateExample {
    private function privateMethod() {
        return 42;
    }
}
$instance = Phake::mock(PrivateExample::class);
Phake::when($instance)->privateMethod()->thenCallParent();
var_dump(Phake::makevisible($instance)->privateMethod());
```

It doesn't work this way when you try it, regardless of whether you use a partial mock or a normal mock:

```
Warning:  call_user_func_array() expects parameter 1 to be a valid callback, cannot
access private method PrivateExample::privateMethod() in
vendor/phake/phake/src/Phake/ClassGenerator/EvalLoader.php(58) : eval()'d code on line 46
```

`makeVisible()` only works on mocks; it can't be used on normal instances:

```
Catchable fatal error: Argument 1 passed to Phake::makeVisible() must be an instance of
Phake_IMock, instance of PrivateExample given, called in test.php on line 12 and defined
in vendor/phake/phake/src/Phake.php on line 660
```

I can see where this feature might be useful with a protected method...

```php
class ProtectedExample {
    protected function protectedMethod() {
        return 42;
    }
}
$instance = Phake::partialMock(ProtectedExample::class);
var_dump(Phake::makevisible($instance)->protectedMethod());
```

Because, in that instance, it does work:

```
int(42)
```

But, if there's a use case in which it works for a `private` method, I'm not seeing it.

I'll be giving a presentation on Phake at [PHP UK](http://phpconference.co.uk/schedule/) on 2/18. If it'd be possible to get clarification on this before then, I'd appreciate it.

**Comment by mlively** [comment]
The documentation here is definitely lacking, I am in the process of cleaning that up now.

Furthermore, there is definitely a bug with this working with private methods (It appears I never tested on private methods, shame on me.) Basically what this does internally is returns a proxy for whatever mock you passed to it. This proxy implements only the __call method and in that method it basically uses reflection to expose the same method in the underlying class. The problem appears to be that in order to modify the visibility of a method, you have to make sure you have the reflection method instance that is tied to the declaring class, not the called class. This is why protected methods work just fine. Those are overridden in the mock object and the mock method can still reach the parent due to the mock extending the class in question. I am very certain I never implement private methods in mocks when the mock is generated, I do believe even trying to do so would result in an error. So the visibility proxy isn't actually making the right method visible.

This also makes the example of setting a stub on the private method really silly, because there is nothing there to stub.

The use case for is actually much more succinctly stated in the phpdocs for the visibility proxy then in the documentation.

    Acts as a proxy to any object that allows calling any private or protected method on the wrapper and
    forward those calls to the wrapped object.
    
    I debated alot about putting anything like this in Phake and at some point I may just pull it out into its
    own library. It should be used with great caution and should really only need to be used to help ease 
    steps in refactoring of otherwise hard to reach private and protected methods

Basically this should really only be used to help in refactoring private or protected (but until I fix it, really only protected) methods that are buried deep in nested calls. So instead of having to figure out how to eventually call that method through the public interface of the class with all of the correct permutations to cover the complexity of the protected or private method itself, you can just mock it, expose it, and call it.

I'm keeping this open to both fix the bug as well as update the documentation to give a realistic example of use. Any thoughts you have on the functionality or ideas for documentation would be welcomed. This should be wrapped up for you today or tomorrow.
